DevTools/HackStudio/LanguageServers/Cpp/main.cpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2020, Itamar S. <itamar8910@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/LexicalPath.h>
+#include <DevTools/HackStudio/LanguageServers/Cpp/ClientConnection.h>
+#include <LibCore/EventLoop.h>
+#include <LibCore/File.h>
+#include <LibCore/LocalServer.h>
+#include <LibIPC/ClientConnection.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+static const char* local_socket_path = "/tmp/cpp-language";
```

Comment:
This makes the server global which seems un-optimal if it's only intended to service a specific instance of HackStudio :)

Reply:
> This makes the server global which seems un-optimal if it's only intended to service a specific instance of HackStudio :)

I would expect multiple HackStudio instances with different projects to work with this architecture? And of course, global seems over-kill, I don't think it should be running at all times.

Reply:
Yeah, one server could technically support connections from multiple instances of HackStudio, but it's definitely a better idea to have a separate instance of the language server per HackStudio instance (e.g if the language server crashes).
I changed the local socket path of the language server to be /tmp/language-servers/cpp/\<pid of hackstudio\>.

Reply:
That path is racy since anybody could guess the PID and create a socket before you, intercepting your connection.
Paging @bugaevc, what do you think is a good approach for this sort of thing? Putting sockets in the home directory? Maybe it would be better to pass a file descriptor when forking off the language server and do the IPC over a pipe instead?

Reply:
> /tmp/language-servers/cpp/<pid of hackstudio>

No no no, please don't do that sort of thing.

Why don't we just use SystemServer MultiInstance services for this again?

Reply:
The only thing I can think of against using SystemServer would be that "this isn't a system-wide server".
imagine having to make a new entry for every language server we add, or the user decides to add.

Reply:
Perhaps we should have a separate userland config file for system server?
it would be rw for some user (or all users?) and HackStudio could just register its language servers with that

Reply:
> The only thing I can think of against using SystemServer would be that "this isn't a system-wide server".

But it is — in the same way that WebContent, ProtocolServer, and Clipboard are.

I get where you're coming from though; it would sure be nice to separate "actually system services" from "whatever the user wants to run on their own" — we could have per-user SystemServer configs, or even per-user SystemServer's, like launchd and systemd do.

But what I want to emphasize is that we've kind of punted on this distinction (system-wide vs user-specific) for a long time, and fixing/untangling that (if we even want to go there) is going to take way more than just the C++ language server. Or in other words, the language server is not any different from the other servers in this regard.

Reply:
> Why don't we just use SystemServer MultiInstance services for this again?

That sounds good yeah, a service with `Lazy` and `MultiInstance` seems to be what we need here.

Having a per-user SystemServer config directory sounds really nice.
I think it would be out of the scope of this PR though, so for now I suggest we go with putting the C++ language server service in the global `SystemServer.ini`.

Reply:
> I think it would be out of the scope of this PR though, so for now I suggest we go with putting the C++ language server service in the global `SystemServer.ini`.

Of course, I wasn't suggesting that needs to be done in the PR (or any time soon).

Reply:
Okay, I pushed the changes. I'm marking this as resolved, but ofc feel free to reopen this conversation.